eth/vm/state.py

```diff
@@ -159,19 +172,42 @@ def state_root(self) -> Hash32:
         return self._account_db.state_root
 
     def make_state_root(self) -> Hash32:
+        new_storage_roots = self._storage_db.make_storage_roots()
+
+        for address, storage_root in new_storage_roots:
+            self.logger.debug2(
+                "Updating account 0x%s to storage root 0x%s",
+                address.hex(),
+                storage_root.hex(),
+            )
+            self._account_db.set_storage_root(address, storage_root)
+
         return self._account_db.make_state_root()
 
     def get_storage(self, address: Address, slot: int, from_journal: bool=True) -> int:
-        return self._account_db.get_storage(address, slot, from_journal)
+        validate_canonical_address(address, title="Storage Address")
+        validate_uint256(slot, title="Storage Slot")
+
+        return self._storage_db.get(address, slot, from_journal)
 
     def set_storage(self, address: Address, slot: int, value: int) -> None:
-        self._account_db.set_storage(address, slot, value)
+        validate_uint256(value, title="Storage Value")
+        validate_uint256(slot, title="Storage Slot")
+        validate_canonical_address(address, title="Storage Address")
+
+        self._storage_db.set(address, slot, value)
 
     def delete_storage(self, address: Address) -> None:
-        self._account_db.delete_storage(address)
+        validate_canonical_address(address, title="Storage Address")
+
+        self._storage_db.delete(address)
+        self._account_db.set_storage_root(address, BLANK_ROOT_HASH)
 
     def delete_account(self, address: Address) -> None:
+        validate_canonical_address(address, title="Storage Address")
+
         self._account_db.delete_account(address)
+        self._storage_db.delete(address)
```

Comment:
I think that this orphans data. I'm not sure but, `_storage_db.delete` doesn't do anything, it reduces to a `JournalDB.clear()` which isn't persisted. So the account is deleted but the state trie is left behind in the database, which would lead to ever growing disk usage.

Reply:
Maybe this also needs a call to `self._account_db.set_storage_root(address, BLANK_ROOT_HASH)`? I'm not sure

Reply:
Pruning old state tries is something we're not attempting to do at this point, for the account or storage tries. Trinity is essentially an archive node at this point.